fix(kitsu-core): fix inability to link relationship links on circular resources

pedep opened the pull request:
Before this fix, this error would occasionally pop up
```
    TypeError: Cannot read properties of undefined (reading 'links')

      65 |   const cache = previouslyLinked[`${resource.type}#${resource.id}`]
      66 |   data[key].data = cache || link(resource, included, previouslyLinked)
    > 67 |   if (data.relationships[key].links) data[key].links = data.relationships[key].links
         |                               ^
      68 |   delete data.relationships[key]
      69 | }
      70 |

      at links (packages/kitsu-core/src/linkRelationships/index.js:67:31)
      at linkObject (packages/kitsu-core/src/linkRelationships/index.js:123:7)
      at deserialiseArray (packages/kitsu-core/src/deserialise/index.js:14:13)
      at deserialiseArray (packages/kitsu-core/src/deserialise/index.js:61:48)
      at Object.<anonymous> (packages/kitsu-core/src/deserialise/index.spec.js:795:42)
```
When serializing a resource with a once removed relationship to itself (A -> B -> A).
The problem seems to stem from same origin as this issue https://github.com/wopian/kitsu/issues/579, where a `relationships` object is deleted before subsequent objects with references to it has had a change to read it

The simple solution would be to add safe navigation to the `if` statement, but then `meta` and `links` would be lost on any subsequent inclusions of the resource.

I was unable to reproduce this issue with array resource linkages, so only `linkObject` was modified. I have a feeling it might be due to something relating to pass-by-value/pass-by-reference. I wrote a spec for it, and it passed without modification, so i just removed it again. I can add it back if you think it would be relevant

I'm very open to style changes or suggestions for alternate implementations

wopian commented on packages/kitsu-core/src/linkRelationships/index.js:
```diff
@@ -57,14 +59,21 @@ function linkArray (data, included, key, previouslyLinked) {
-  if (data.relationships[key].links) data[key].links = data.relationships[key].links
+  data[key].data = cache || link(resource, included, previouslyLinked, relationshipCache)
+
+  const relationships = relationshipCache[`${data.type}#${data.id}#${key}`] || data.relationships[key]
+  if (!relationshipCache[`${data.type}#${data.id}#${key}`]) relationshipCache[`${data.type}#${data.id}#${key}`] = relationships
```
Replace repeated usage of `${data.type}#${data.id}#${key}` with:

```js
const cacheKey = `${data.type}#${data.id}#${key}`
```

```suggestion
  const cacheKey = `${data.type}#${data.id}#${key}`
  const relationships = relationshipCache[cacheKey] || data.relationships[key]
  if (!relationshipCache[cacheKey]) relationshipCache[cacheKey] = relationships
```

wopian commented on packages/kitsu-core/src/linkRelationships/index.js:
```diff
@@ -57,14 +59,21 @@ function linkArray (data, included, key, previouslyLinked) {
+  const relationships = relationshipCache[`${data.type}#${data.id}#${key}`] || data.relationships[key]
+  if (!relationshipCache[`${data.type}#${data.id}#${key}`]) relationshipCache[`${data.type}#${data.id}#${key}`] = relationships
+
+  if (relationships && relationships.links) data[key].links = relationships.links
```
Can be simplified with optional chaining

```suggestion
  if (relationships?.links) data[key].links = relationships.links
```

wopian commented on packages/kitsu-core/src/linkRelationships/index.js:
```diff
@@ -57,14 +59,21 @@ function linkArray (data, included, key, previouslyLinked) {
+  if (!relationshipCache[`${data.type}#${data.id}#${key}`]) relationshipCache[`${data.type}#${data.id}#${key}`] = relationships
+
+  if (relationships && relationships.links) data[key].links = relationships.links
+  if (relationships && relationships.meta) data[key].meta = relationships.meta
```
```suggestion
  if (relationships?.meta) data[key].meta = relationships.meta
```

wopian commented:
> I was unable to reproduce this issue with array resource linkages, so only linkObject was modified. I have a feeling it might be due to something relating to pass-by-value/pass-by-reference. I wrote a spec for it, and it passed without modification, so i just removed it again. I can add it back if you think it would be relevant

Even though it passed as-is, I think it's a good idea to keep the test as it's extra redundency to protect against regressions

pedep commented:
@wopian Turns out i spoke too soon regarding the array case.
I added back in the `meta` keys to the relationship objects after i deleted the spec. Turns out `isArray` does not handle `meta` at the top level, so i handled this in 0d5577960df833db0ec136dd60f4321495e63ac1

I'll go squash the commits in preparation for the merge when you give the OK

wopian commented:
You don't need to squash the commits, the PR merge itself squashes them 👍

wopian commented:
Thank you! 🎉

wopian commented:
Released in [10.0.0-alpha.26](https://github.com/wopian/kitsu/releases/tag/v10.0.0-alpha.26)
